homeassistant/components/media_player/directv.py

```diff
@@ -47,73 +55,287 @@
+
+        directv_entity = {
+            CONF_NAME: name,
+            CONF_HOST: host,
+            CONF_PORT: port,
+            CONF_DEVICE: device,
+            RECEIVER_ID: resp.get('receiverId', host).replace(' ', ''),
+            CONF_DISCOVER_CLIENTS: config.get(CONF_DISCOVER_CLIENTS),
+            CONF_CLIENT_DISCOVER_INTERVAL: client_discover_interval,
+        }
 
     elif discovery_info:
+        discovered = True
         host = discovery_info.get('host')
-        name = 'DirecTV_{}'.format(discovery_info.get('serial', ''))
-
-        # Attempt to discover additional RVU units
-        _LOGGER.debug("Doing discovery of DirecTV devices on %s", host)
+        serial_num = discovery_info.get('serial', host)
+        receiver_id = discovery_info.get('serial', '').split('-')[-1]
+        receiver_id = receiver_id if receiver_id else host
+        name = 'DirecTV_{}'.format(serial_num)
+        resp = []
+
+        if (receiver_id, DEFAULT_DEVICE) in known_devices:
+            _LOGGER.debug("Discovered device on host %s is already"
+                          " configured", host)
+            return
 
         from DirectPy import DIRECTV
-        dtv = DIRECTV(host, DEFAULT_PORT)
         try:
-            resp = dtv.get_locations()
+            dtv = DIRECTV(host, DEFAULT_PORT, DEFAULT_DEVICE)
         except requests.exceptions.RequestException as ex:
-            # Bail out and just go forward with uPnP data
-            # Make sure that this device is not already configured
-            # Comparing based on host (IP) and clientAddr.
-            _LOGGER.debug("Request exception %s trying to get locations", ex)
-            resp = {
-                'locations': [{
-                    'locationName': name,
-                    'clientAddr': DEFAULT_DEVICE
-                }]
-            }
-
-        _LOGGER.debug("Known devices: %s", known_devices)
+            _LOGGER.debug("Request exception %s trying to connect to %s",
+                          ex, name)
+            resp = []
+
+        if dtv:
+            try:
+                resp = dtv.get_locations()
+            except requests.exceptions.RequestException as ex:
+                _LOGGER.debug("Request exception %s trying to get "
+                              "locations for %s", ex, name)
+                resp = []
+
         for loc in resp.get("locations") or []:
-            if "locationName" not in loc or "clientAddr" not in loc:
+            if loc.get("clientAddr") == DEFAULT_DEVICE and \
+               "locationName" in loc:
+                name = str.title(loc["locationName"])
+                break
+
+        _LOGGER.debug("Adding discovered device %s on host %s",
+                      name, host)
+        directv_entity = {
+            CONF_NAME: name,
+            CONF_HOST: host,
+            CONF_PORT: DEFAULT_PORT,
+            CONF_DEVICE: DEFAULT_DEVICE,
+            RECEIVER_ID: receiver_id,
```

Comment:
In the other case you replaced spaces with empty strings for receiver id.

Reply:
The difference is how the serial number is received. When the serial number is retrieved through the API then it comes in as 3 groups of 4 numbers each separated by a space: 1234 5678 9012
Through the HASS discovery, we also get the serial number but then it comes in as: DirecTV-123456789012

Hence, when getting it through the API we remove the spaces resulting in: 123456789012
But when serial is from the discovery then instead we split on - and grab the second (or last) entry resulting in: 123456789012